File: go-backend/handlers/task_statuses.go
package handlers

import (
	"encoding/json"
	"go-backend/models"
	"go-backend/services"
	"log"
	"net/http"
	"strconv"

	"github.com/gorilla/mux"
)

// GetTaskStatusesRoute retrieves all task statuses for the current user
func (s *Handler) GetTaskStatusesRoute(w http.ResponseWriter, r *http.Request) {
	userID := r.Context().Value("current_user").(int)

	statuses, err := services.GetTaskStatuses(s.GetDB(), userID)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(statuses)
}

// GetTaskStatusRoute retrieves a single task status by ID
func (s *Handler) GetTaskStatusRoute(w http.ResponseWriter, r *http.Request) {
	userID := r.Context().Value("current_user").(int)
	statusID, err := strconv.Atoi(mux.Vars(r)["id"])
	if err != nil {
		http.Error(w, "Invalid status ID", http.StatusBadRequest)
		return
	}

	status, err := services.GetTaskStatus(s.GetDB(), userID, statusID)
	if err != nil {
		http.Error(w, "Task status not found", http.StatusNotFound)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(status)
}

// CreateTaskStatusRoute creates a new task status
func (s *Handler) CreateTaskStatusRoute(w http.ResponseWriter, r *http.Request) {
	userID := r.Context().Value("current_user").(int)

	var params models.CreateTaskStatusParams
	if err := json.NewDecoder(r.Body).Decode(&params); err != nil {
		http.Error(w, "Invalid request payload", http.StatusBadRequest)
		return
	}

	// Validate required fields
	if params.Name == "" || params.DisplayName == "" || params.Color == "" {
		http.Error(w, "name, display_name, and color are required", http.StatusBadRequest)
		return
	}

	statusID, err := services.CreateTaskStatus(s.GetDB(), userID, params)
	if err != nil {
		log.Printf("Error creating task status: %v", err)
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	w.WriteHeader(http.StatusCreated)
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]interface{}{"id": statusID})
}

// UpdateTaskStatusRoute updates an existing task status
func (s *Handler) UpdateTaskStatusRoute(w http.ResponseWriter, r *http.Request) {
	userID := r.Context().Value("current_user").(int)
	statusID, err := strconv.Atoi(mux.Vars(r)["id"])
	if err != nil {
		http.Error(w, "Invalid status ID", http.StatusBadRequest)
		return
	}

	var params models.UpdateTaskStatusParams
	if err := json.NewDecoder(r.Body).Decode(&params); err != nil {
		http.Error(w, "Invalid request payload", http.StatusBadRequest)
		return
	}

	err = services.UpdateTaskStatus(s.GetDB(), userID, statusID, params)
	if err != nil {
		log.Printf("Error updating task status: %v", err)
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	w.WriteHeader(http.StatusOK)
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(models.GenericResponse{
		Message: "success",
		Error:   false,
	})
}

// DeleteTaskStatusRoute deletes a task status
func (s *Handler) DeleteTaskStatusRoute(w http.ResponseWriter, r *http.Request) {
	userID := r.Context().Value("current_user").(int)
	statusID, err := strconv.Atoi(mux.Vars(r)["id"])
	if err != nil {
		http.Error(w, "Invalid status ID", http.StatusBadRequest)
		return
	}

	err = services.DeleteTaskStatus(s.GetDB(), userID, statusID)
	if err != nil {
		log.Printf("Error deleting task status: %v", err)
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	w.WriteHeader(http.StatusNoContent)
}

// ReorderTaskStatusesRoute reorders task statuses
func (s *Handler) ReorderTaskStatusesRoute(w http.ResponseWriter, r *http.Request) {
	userID := r.Context().Value("current_user").(int)

	var params models.ReorderTaskStatusesParams
	if err := json.NewDecoder(r.Body).Decode(&params); err != nil {
		http.Error(w, "Invalid request payload", http.StatusBadRequest)
		return
	}

	if len(params.StatusIDs) == 0 {
		http.Error(w, "status_ids array is required", http.StatusBadRequest)
		return
	}

	err := services.ReorderTaskStatuses(s.DB, userID, params.StatusIDs)
	if err != nil {
		log.Printf("Error reordering task statuses: %v", err)
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	w.WriteHeader(http.StatusOK)
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(models.GenericResponse{
		Message: "success",
		Error:   false,
	})
}

File: zettelkasten-front/src/api/taskStatuses.ts
import {
  TaskStatus,
  CreateTaskStatusParams,
  UpdateTaskStatusParams,
  ReorderTaskStatusesParams,
} from '../models/TaskStatus';

const base_url = import.meta.env.VITE_URL;

export async function fetchTaskStatuses(): Promise<TaskStatus[]> {
  const response = await fetch(`${base_url}/task-statuses`, {
    headers: {
      Authorization: `Bearer ${localStorage.getItem('token')}`,
    },
  });

  if (!response.ok) {
    throw new Error('Failed to fetch task statuses');
  }

  return await response.json();
}

export async function fetchTaskStatus(id: number): Promise<TaskStatus> {
  const response = await fetch(`${base_url}/task-statuses/${id}`, {
    headers: {
      Authorization: `Bearer ${localStorage.getItem('token')}`,
    },
  });

  if (!response.ok) {
    throw new Error('Failed to fetch task status');
  }

  return await response.json();
}

export async function createTaskStatus(
  params: CreateTaskStatusParams,
): Promise<{ id: number }> {
  const response = await fetch(`${base_url}/task-statuses`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${localStorage.getItem('token')}`,
    },
    body: JSON.stringify(params),
  });

  if (!response.ok) {
    throw new Error('Failed to create task status');
  }

  return await response.json();
}

export async function updateTaskStatus(
  id: number,
  params: UpdateTaskStatusParams,
): Promise<{ message: string; error: boolean }> {
  const response = await fetch(`${base_url}/task-statuses/${id}`, {
    method: 'PUT',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${localStorage.getItem('token')}`,
    },
    body: JSON.stringify(params),
  });

  if (!response.ok) {
    throw new Error('Failed to update task status');
  }

  return await response.json();
}

export async function deleteTaskStatus(id: number): Promise<void> {
  const response = await fetch(`${base_url}/task-statuses/${id}`, {
    method: 'DELETE',
    headers: {
      Authorization: `Bearer ${localStorage.getItem('token')}`,
    },
  });

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(errorText || 'Failed to delete task status');
  }
}

export async function reorderTaskStatuses(
  params: ReorderTaskStatusesParams,
): Promise<{ message: string; error: boolean }> {
  const response = await fetch(`${base_url}/task-statuses/reorder`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${localStorage.getItem('token')}`,
    },
    body: JSON.stringify(params),
  });

  if (!response.ok) {
    throw new Error('Failed to reorder task statuses');
  }

  return await response.json();
}
